com.chartboost.helium/Android~/helium-android-bridge/src/main/java/com/chartboost/heliumsdk/unity/HeliumUnityAdWrapper.kt

```diff
@@ -0,0 +1,225 @@
+        return ad as HeliumBannerAd?
+    }
+
+    private fun asFullScreen(): HeliumFullscreenAd? {
+        return ad as HeliumFullscreenAd?
+    }
+
+    fun load() {
+        HeliumUnityBridge.runTaskOnUiThread {
+            ad.load()
+            startedLoad = true
+        }
+    }
+
+    fun load(screenLocation: Int) {
+        HeliumUnityBridge.runTaskOnUiThread {
+            createBannerLayout(screenLocation)
+            load()
+        }
+    }
+
+    fun show() {
+        HeliumUnityBridge.runTaskOnUiThread {
+            if (ad is HeliumFullscreenAd)
+                asFullScreen()?.show()
+        }
+    }
+
+    fun setKeyword(keyword: String, value: String): Boolean {
+        return ad.keywords.set(keyword, value)
+    }
+
+    fun removeKeyword(keyword: String): String {
+        return ad.keywords.remove(keyword) ?: ""
+    }
+
+    fun setCustomData(customData: String) {
+        if (ad is HeliumRewardedAd) {
+            ad.customData = customData
+        } else {
+            Log.d(TAG,"custom data can only be set on a rewarded ad")
+        }
+    }
+
+    fun setBannerVisibility(isVisible: Boolean) {
+        HeliumUnityBridge.runTaskOnUiThread {
+            if (ad is HeliumBannerAd && bannerLayout != null) {
+                val visibility = if (isVisible) View.VISIBLE else View.INVISIBLE
+                bannerLayout?.visibility = visibility
+                asBanner()?.visibility = visibility
+            }
+        }
+    }
+
+    fun clearLoaded(): Boolean {
+        if (ad is HeliumFullscreenAd)
+            return asFullScreen()?.clearLoaded() ?: false
+        if (ad is HeliumBannerAd) {
+            HeliumUnityBridge.runTaskOnUiThread { asBanner()?.clearAd() }
+            return true
+        }
+        return false
+    }
+
+    fun readyToShow(): Boolean {
+        return if (!startedLoad) false else try {
+            when (ad) {
+                is HeliumFullscreenAd -> asFullScreen()?.readyToShow() ?: false
+                is HeliumBannerAd -> {
+                    Log.w(TAG, "This should never be called, banners do not have readyToShow")
+                    false
+                }
+                else -> false
+            }
+        } catch (ex: Exception) {
+            Log.w(TAG, "Helium encountered an error calling Ad.readyToShow() - ${ex.message}")
+            false
+        }
+    }
+
+    fun destroy() {
+        HeliumUnityBridge.runTaskOnUiThread {
+            destroyBannerLayout()
+            ad.destroy()
+        }
+    }
+
+    private fun createBannerLayout(screenLocation: Int) {
+        if (ad !is HeliumBannerAd)
+            return
+
+        bannerLayout?.let {
+            it.removeAllViews()
+        }
+
+        // Create the banner layout on the given position.
+        // Check if there is an already existing banner layout. If so, remove it. Otherwise,
+        // create a new one.
+        bannerLayout?.let {
+            val bannerParent = it.parent as FrameLayout
```

Comment:
you might want to consider using ViewGroup here as how do you know it's a FrameLayout? Or is this a Unity thing?

Reply:
This is a lot of legacy existing code that I have not written myself. For now, I will keep it this way unless strictly necessary, I do not want to make modifications yet that go beyond the scope of the conversion to Kotlin. I can def give it a try later

Reply:
Actually, giving it a try right now

Reply:
After doing changes to this, Banners seem to work better :) thanks